src/group_theory/dihedral.lean

```diff
@@ -0,0 +1,178 @@
+/-
+Copyright (c) 2020 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+import data.zmod.basic
+import group_theory.order_of_element
+
+/-!
+# Dihedral Groups
+
+We define the dihedral groups `dihedral n`, with elements `r i` and `sr i` for `i : zmod n`.
+
+For `n ≠ 0`, `dihedral n` represents the symmetry group of the regular `n`-gon. `r i` represents
+the rotations of the `n`-gon by `2πi/n`, and `sr i` represents the reflections of the `n`-gon.
+`dihedral 0` corresponds to the infinite dihedral group.
+-/
+
+/--
+For `n ≠ 0`, `dihedral n` represents the symmetry group of the regular `n`-gon. `r i` represents
+the rotations of the `n`-gon by `2πi/n`, and `sr i` represents the reflections of the `n`-gon.
+`dihedral 0` corresponds to the infinite dihedral group.
+-/
+@[derive decidable_eq]
+inductive dihedral (n : ℕ) : Type
+| r : zmod n → dihedral
+| sr : zmod n → dihedral
+
+namespace dihedral
+
+variables {n : ℕ}
+
+/--
+Multiplication of the dihedral group.
+-/
+def mul : dihedral n → dihedral n → dihedral n
+| (r i) (r j) := r (i + j)
+| (r i) (sr j) := sr (j - i)
+| (sr i) (r j) := sr (i + j)
+| (sr i) (sr j) := r (j - i)
+
+/--
+The identity `1` is the rotation by `0`.
+-/
+def one : dihedral n := r 0
+
+instance : inhabited (dihedral n) := ⟨one⟩
+
+/--
+The inverse of a an element of the dihedral group.
+-/
+def inv : dihedral n → dihedral n
+| (r i) := r (-i)
+| (sr i) := sr i
```

Comment:
It's tempting to mark these `private`, to encourage users to use the `group` operators instead of these names.

Reply:
I think it's fine to mark such defs `private`.